Type Parameter Refactoring

b-butler opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should based on *maint*. -->
<!-- New features should based on *master*. -->

## Description

This PR fixes #1068 and does some maintenance improvements to the type parameter implementation and documentation. While this does improve both documentation and implementation, this is not intended to be the final refactoring support for in-place mutation of nested data structures is still necessary, and better rearrangement of classes in `hoomd/data` should be done.

In addition, developer documentation on this infrastructure via `ARCITECTURE.md` is still missing.
<!-- Describe your changes in detail. -->

## Motivation and context

This fixes #1068 and addresses some other needs for the type parameter API/internals.

## How has this been tested?

Existing tests pass.
<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Improved documentation for ``hoomd.data.typeparam.TypeParameter``.
Fixed setting type pairs with ``TypeParameter`` instances after calling ``hoomd.Simulation.run`` (#1068).
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

joaander commented on hoomd/data/typeparam.py:
```diff
@@ -4,12 +4,87 @@
+    is disallowed).*
+
+    `TypeParameter` instances extend the base Python mapping interface with with
+    smart defaults, value/key validation/processing, and advanced indexing.
```
Explain that the class store parameters indexed by type name or pairs of type names.

joaander commented on hoomd/data/typeparam.py:
```diff
@@ -4,12 +4,87 @@
+        print(lj.params[("A", "B")])
+        # {"epsilon": 4.0, "sigma": 1.0}
+
+    Note:
+        Setting values for fictitious but valid types will not trigger an error
+        before calling `hoomd.Simulation.run`, but attempts to access that data
+        or set values for fictitious types afterwards will result in
+        ``KeyError`` exceptions.
```
I would reword this in a more positive way. Something covering these points.
* Before calling `run`, `TypeParameterDict` can map parameters for any number of arbitrarily named types.
* After calling run, it contains only those types that are present in the simulation `State`.
* Previously set names that are not in `State` are lost.
* It is an error to set names not present in `State` after calling run.

joaander commented on hoomd/data/typeparam.py:
```diff
@@ -4,12 +4,87 @@
+        ``KeyError`` exceptions.
+
+    Warning:
+        For nested data structures, editing the internal stuctures such as a
```
Start off with something positive like "Values in this mapping are returned by copy, not reference. To modify nested data structures, such as the union shape intergrators in HPMC and the table pair potential in MD, read the current value, modify it, then write a new parameter (example)."

Also, don't users "anti-pattern". Most users won't know what that means.

joaander commented on hoomd/data/typeparam.py:
```diff
@@ -4,12 +4,87 @@
+class TypeParameter(MutableMapping):
+    """Implement a type based mutable mapping.
+
+    *Implements the `collections.abc.MutableMapping` interface (``__delitem__``
```
This doesn't appear to render correctly in the docs.

joaander commented on hoomd/data/typeparam.py:
```diff
@@ -4,12 +4,87 @@
+
+    .. rubric:: Indexing
+
+    For getting and setting values, multiple index formats are supported. The
```
Explain basic indexing before advanced indexing.

joaander commented on hoomd/data/typeparam.py:
```diff
@@ -34,6 +109,40 @@ def __setitem__(self, key, value):
+        Args:
+            key:
+                Valid keys specifications (depends on the expected key length).
+            default (``any``, optional):
```
`default` does not appear in the argument list for this method.

joaander commented on hoomd/data/typeparam.py:
```diff
@@ -34,6 +109,40 @@ def __setitem__(self, key, value):
+
+        Returns:
+            values:
+                Returns a dict of the values for the keys asked for if multiple
```
This can handle a list of keys? 

Or is this referring to ('A', 'B') "keys"?

b-butler replied:
All allowed methods can handle the full indexing documented above in a mostly intuitive way (still no deleting of items allowed).

joaander reviewed (changes requested):
I have some suggestions to improve the documentation.

joaander reviewed (commented):
Thanks, this is looking better. It needs to pass tests on CI before approval.

b-butler commented:
@joaander, yes they do. I added a type to the dummy state, and then checked the tests, I modified, and it ended up causing errors in other tests I didn't run :sweat_smile:. I ran all the tests and it seems to work now :crossed_fingers:

joaander reviewed (approved):
Thanks!
